### PR TITLE
use free subnet to create network

### PR DESCRIPTION
**Pull request by junnplus**
Signed-off-by: Ye Sijun <junnplus@gmail.com>

We use increment id as the subnet part in the past, it is easy to cause subnet conflicts.

Introduce `GetFreeSubnet` to find a free subnet, and check conflict when specify a subnet

### CODE REVIEW

File: pkg/netutil/netutil.go [hunk 1]
```diff
@@ -43,15 +47,17 @@ func NewCNIEnv(cniPath, cniConfPath string) (*CNIEnv, error) {
...
-	err = e.ensureDefaultNetworkConfig()
-	if err != nil {
+	if err := os.MkdirAll(e.NetconfPath, 0755); err != nil {
+		return nil, err
+	}
+	if err := e.ensureDefaultNetworkConfig(); err != nil {
 		return nil, err
 	}
-	e.Networks, err = e.networkConfigList()
+	networks, err := e.networkConfigList()
```
[the remark this answers is not in the document]
Comment (junnplus) [reply pkg/netutil/netutil.go:56]: Not why, both are ok.

File: pkg/netutil/subnet/subnet.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+*/
+
+package subnet
+
+import (
+	"fmt"
+	"net"
+)
+
+func GetLiveNetworkSubnets() ([]*net.IPNet, error) {
```
[the remark this answers is not in the document]
Comment (junnplus) [reply pkg/netutil/subnet/subnet.go:24]: We can't be sure what subnets exist. In addition, the new network interface requires root. happy case no point

File: pkg/netutil/netutil.go [hunk 3]
```diff
@@ -63,25 +69,111 @@ func (e *CNIEnv) NetworkMap() map[string]*networkConfig {
...
+
+func (e *CNIEnv) RemoveNetwork(net *networkConfig) error {
+	fn := func() error {
+		if err := os.RemoveAll(net.File); err != nil {
+			return err
+		}
+		if err := net.clean(); err != nil {
+			return err
+		}
+		e.Networks, _ = e.networkConfigList()
```
[the remark this answers is not in the document]
Comment (junnplus) [reply pkg/netutil/netutil.go]: Remove it first, not required.

File: pkg/netutil/netutil.go [hunk 4]
```diff
@@ -117,11 +209,8 @@ func (e *CNIEnv) GenerateNetworkConfig(labels []string, id int, name string, plu
 	}, nil
 }
 
-// WriteNetworkConfig writes networkConfig file to cni config path.
-func (e *CNIEnv) WriteNetworkConfig(net *networkConfig) error {
-	if err := os.MkdirAll(e.NetconfPath, 0755); err != nil {
-		return err
-	}
+// writeNetworkConfig writes networkConfig file to cni config path.
+func (e *CNIEnv) writeNetworkConfig(net *networkConfig) error {
```
Comment (fuweid) [inline pkg/netutil/netutil.go:213]: Both the Ctrl-C or No space can cancel the write and maybe there is invalid data in file. Suggest to use atomic write and make sure that content is correct.

File: pkg/netutil/netutil.go [hunk 5]
```diff
@@ -171,76 +246,75 @@ func (e *CNIEnv) networkConfigList() ([]*networkConfig, error) {
...
+func (e *CNIEnv) parseSubnet(subnetStr string) (*net.IPNet, error) {
+	usedSubnets, err := e.usedSubnets()
 	if err != nil {
-		return nil, fmt.Errorf("failed to parse subnet %q", subnetStr)
+		return nil, err
 	}
-	if !subnet.IP.Equal(subnetIP) {
-		return nil, fmt.Errorf("unexpected subnet %q, maybe you meant %q?", subnetStr, subnet.String())
+	var subnet *net.IPNet
+	if subnetStr == "" {
```
Comment (fuweid) [inline pkg/netutil/netutil.go:283]: Suggest: this branch can be moved to the beginning of function and fast return instead of using else branch~

Comment (junnplus) [reply pkg/netutil/netutil.go:283]: I don't get it. Can you get me more explain?

File: pkg/netutil/netutil.go (hunk 4 above)
Comment (junnplus) [reply pkg/netutil/netutil.go:213]: If needed, maybe another PR

File: pkg/netutil/netutil.go (hunk 5 above)
Comment (junnplus) [reply pkg/netutil/netutil.go:282]: I got it.

File: pkg/netutil/subnet/subnet.go [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+		if err != nil {
+			return nil, err
+		}
+		nets = append(nets, n)
+	}
+	return nets, nil
+}
+
+// GetFreeSubnet try to find a free subnet in the given network
+func GetFreeSubnet(n *net.IPNet, usedNetworks []*net.IPNet) (*net.IPNet, error) {
```
Comment (AkihiroSuda) [inline pkg/netutil/subnet/subnet.go:41]: This should need a lock file to support running multiple `nerdctl network create` concurrently?

Comment (junnplus) [reply pkg/netutil/subnet/subnet.go:41]: Lock cni path dir when create https://github.com/containerd/nerdctl/pull/1245/files#diff-669913e187d4dc726dcb9e01a5666ae0562549545138d25d2c8fc2ceb3e23d30R134

**Review by fuweid — APPROVED** [review]

**Comment by AkihiroSuda** [comment]
Seems to have caused a regression
- #1490

**Pull request merged**